src/coreclr/vm/callstubgenerator.cpp

```diff
@@ -0,0 +1,1013 @@
+extern "C" void CallJittedMethodRetI8Double(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRetDoubleI8(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRetDoubleDouble(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+#endif
+
+#ifdef TARGET_ARM64
+extern "C" void CallJittedMethodRet2I8(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRet2Double(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRet3Double(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRet4Double(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRetFloat(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRet2Float(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRet3Float(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+extern "C" void CallJittedMethodRet4Float(PCODE *routines, int8_t*pArgs, int8_t*pRet, int totalStackSize);
+#endif // TARGET_ARM64
+
+// Generate the call stub for the given method.
+// The returned call stub header must be freed by the caller using FreeCallStub.
+CallStubHeader *CallStubGenerator::GenerateCallStub(MethodDesc *pMD)
+{
+    STANDARD_VM_CONTRACT;
+
+    _ASSERTE(pMD != NULL);
+
+    MetaSig sig(pMD);
+    ArgIterator argIt(&sig);
+
+    m_r1 = NoRange; // indicates that there is no active range of general purpose registers
+    m_r2 = 0;
+    m_x1 = NoRange; // indicates that there is no active range of FP registers
+    m_x2 = 0;
+    m_s1 = NoRange; // indicates that there is no active range of stack arguments
+    m_s2 = 0;
+    m_routineIndex = 0;
+    m_totalStackSize = 0;
+
+    int numArgs = sig.NumFixedArgs() + (sig.HasThis() ? 1 : 0);
+
+    if (argIt.HasThis())
+    {
+        // The "this" argument register is not enumerated by the arg iterator, so
+        // we need to "inject" it here.
+#if defined(TARGET_WINDOWS) && defined(TARGET_AMD64)
+        if (argIt.HasRetBuffArg())
+        {
+            // The return buffer on Windows AMD64 is passed in the first argument register, so the
+            // "this" argument is be passed in the second argument register.
+            m_r1 = 1;
+        }
+        else
+#endif // TARGET_WINDOWS && TARGET_AMD64
+        {
+            // The "this" pointer is passed in the first argument register.
+            m_r1 = 0;
+        }
+    }
+
+    // Allocate space for the routines. The size of the array is conservatively set to twice the number of arguments
+    // plus one slot for the target pointer and reallocated to the real size at the end.
+
+    m_pHeader = (CallStubHeader*)malloc(sizeof(CallStubHeader) + (2 * numArgs + 1) * sizeof(PCODE));
+    if (m_pHeader == NULL)
```

Comment:
Is there are reason to use manual error handling? The method contract is `STANDARD_VM_CONTRACT`, so I would expect it to throw on error.